src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryLogicManager.CreateProvider.cs

```diff
@@ -0,0 +1,334 @@
+            }
+
+            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> Unable to resolve a valid provider; Returns `null`.", s_typeName, methodName);
+            return null;
+        }
+
+        private static object CreateInstance(Type type, string retryMethodName, SqlRetryLogicOption option)
+        {
+            string methodName = MethodBase.GetCurrentMethod().Name;
+            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> Entry point.", s_typeName, methodName);
+
+            if (type == typeof(SqlConfigurableRetryFactory) || type == null)
+            {
+                SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> The given type `{2}` infers as internal `{3}` type."
+                                                      , s_typeName, methodName, type?.Name ?? NullConst, typeof(SqlConfigurableRetryFactory).FullName);
+                MethodInfo internalMethod = typeof(SqlConfigurableRetryFactory).GetMethod(retryMethodName);
+                if (internalMethod == null)
+                {
+                    throw new InvalidOperationException($"Failed to resolve the '{retryMethodName}' method from `{typeof(SqlConfigurableRetryFactory).FullName}` type.");
+                }
+
+                SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> The `{2}.{3}()` method is discovered by `{4}` method name."
+                                                      , s_typeName, methodName, internalMethod.ReflectedType.FullName, internalMethod.Name, retryMethodName);
+                object[] internalFuncParams = PrepareParamValues(internalMethod.GetParameters(), option, retryMethodName);
+                SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> Parameters are prepared to invoke the `{2}.{3}()` method."
+                                                      , s_typeName, methodName, internalMethod.ReflectedType.FullName, internalMethod.Name);
+                return internalMethod.Invoke(null, internalFuncParams);
+            }
+
+            // Searches for the public MethodInfo from the specified type by the given method name
+            // It is a case-sensitive search
+            MethodInfo method = type.GetMethod(retryMethodName);
+            if (method == null)
+            {
+                throw new InvalidOperationException($"Failed to resolve the '{retryMethodName}' method from `{type.FullName}` type.");
+            }
+            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> The `{2}` method metadata is extracted from the `{3}` type by `{4}` method name."
+                                                    , s_typeName, methodName, method.Name, type.FullName, retryMethodName);
+
+            if (!typeof(SqlRetryLogicBaseProvider).IsAssignableFrom(method.ReturnType))
+            {
+                throw new InvalidCastException($"Invalid return type; It must be as `{typeof(SqlRetryLogicBaseProvider).FullName}` type.");
```

Comment:
All exception messages need to come from Resource Builder (Strings.resx) and preferably built via SqlUtil.